File: src/main/java/com/corn/data/entity/ServiceStatus.java
package com.corn.data.entity;

import com.corn.data.dto.ServiceComponentDTO;
import com.corn.data.dto.ServiceEventDTO;
import com.corn.data.dto.ServiceStatusDTO;

import javax.persistence.*;
import java.time.Instant;
import java.util.List;
import java.util.Objects;

import static javax.persistence.GenerationType.SEQUENCE;

/**
 * ServiceEvent entity
 *
 * @author Oleg Zaidullin
 */
@SuppressWarnings("unused")
@Entity
@Table(name = "SERVICE_STATUS")
public class ServiceStatus {

    private Long id;
    private String statusString;
    private String statusType;
    private String description;
    private ServiceEvent currentEvent;
    private boolean current;
    private Instant updatedAt;
    private String updatedBy;

    public ServiceStatus() {}

    public ServiceStatus(String statusString,
                         String statusType,
                         String description,
                         ServiceEvent currentEvent,
                         boolean current,
                         Instant updatedAt,
                         String updatedBy) {
        this.statusString = statusString;
        this.statusType = statusType;
        this.description = description;
        this.currentEvent = currentEvent;
        this.current = current;
        this.updatedAt = updatedAt;
        this.updatedBy = updatedBy;
    }

    public static ServiceStatus getNormal(Instant updatedAt, String updatedBy) {
        return new ServiceStatus(
                "Service is operational",
                "normal",
                "Welcome to the Service Status Page. There you can see current information of the service performance. You can bookmark or subscribe to this page for the latest updates.",
                null,
                true,
                updatedAt,
                updatedBy
        );
    }

    @Id
    @SequenceGenerator(name = "SERVICE_STATUS_SEQ", sequenceName = "SEQ_SERVICE_STATUS", allocationSize=1)
    @GeneratedValue(strategy = SEQUENCE, generator = "SERVICE_STATUS_SEQ")
    @Column(name = "ID", nullable = false)
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @Column(name="STATUS_STRING", nullable = false)
    public String getStatusString() {
        return statusString;
    }

    public void setStatusString(String statusString) {
        this.statusString = statusString;
    }

    @Column(name="STATUS_TYPE", nullable = false)
    public String getStatusType() {
        return statusType;
    }

    public void setStatusType(String statusType) {
        this.statusType = statusType;
    }

    @Column(name="DESCRIPTION", nullable = false)
    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    @Column(name = "UPDATED_AT")
    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }

    @Column(name = "UPDATED_BY")
    public String getUpdatedBy() {
        return updatedBy;
    }

    public void setUpdatedBy(String updatedBy) {
        this.updatedBy = updatedBy;
    }


    @OneToOne(cascade = {CascadeType.ALL})
    @JoinColumn(name = "EVENT_ID")
    public ServiceEvent getCurrentEvent() {
        return currentEvent;
    }

    public void setCurrentEvent(ServiceEvent currentEvent) {
        this.currentEvent = currentEvent;
    }

    @Column(name="CURRENT_STATUS", nullable = false)
    public boolean isCurrent() {
        return current;
    }

    public void setCurrent(boolean current) {
        this.current = current;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ServiceStatus that = (ServiceStatus) o;
        return current == that.current &&
                Objects.equals(id, that.id) &&
                Objects.equals(statusString, that.statusString) &&
                Objects.equals(statusType, that.statusType) &&
                Objects.equals(description, that.description) &&
                Objects.equals(currentEvent, that.currentEvent) &&
                Objects.equals(updatedAt, that.updatedAt) &&
                Objects.equals(updatedBy, that.updatedBy);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, statusString, statusType, description, currentEvent, current, updatedAt, updatedBy);
    }

    @Transient
    public ServiceStatusDTO toValue(List<ServiceComponentDTO> components, List<ServiceEventDTO> history) {
        return new ServiceStatusDTO(
                id,
                statusString,
                statusType,
                description,
                currentEvent == null?null:currentEvent.toValue(),
                current,
                updatedAt,
                updatedBy,
                components,
                history
        );
    }
}

File: src/main/frontend/src/app/model/serviceStatus.ts
import {ServiceEvent} from './serviceEvent';
import {ServiceComponent} from './serviceComponent';
import {toLongString} from '../util/dateTime';
import {EventUpdate} from './eventUpdate';

const ACTIVE_STATES: string [] = ['partial', 'accident', 'maintenance'];

export class ServiceStatus {


    constructor(components: ServiceComponent[]) {
        this.components = components;
    }
    id: number;
    MAX_HISTORY_SIZE = 5;
    date: Date = new Date();
    statusString = '';
    description = '';
    statusType = '';
    currentEvent: ServiceEvent = new ServiceEvent();

    components: ServiceComponent [] = [];

    history: ServiceEvent [] = [];

    private static getNormalComponents(): ServiceComponent [] {
        return [
            new ServiceComponent('Tax'),
            new ServiceComponent('Credit Card Processing'),
            new ServiceComponent('Vendor Gateway'),
            new ServiceComponent('PayPal Gateway'),
            new ServiceComponent('Merchant Care'),
            new ServiceComponent('Self Care'),
        ];
    }

    static fromBackend(status): ServiceStatus {
        const serviceStatus = new ServiceStatus(status.components);
        serviceStatus.id = status.id;
        serviceStatus.statusType = status.statusType;
        serviceStatus.statusString = status.statusString;
        serviceStatus.description = status.description;
        serviceStatus.currentEvent = status.currentEvent ? ServiceEvent.fromBackend(status.currentEvent, status.components) : null;
        serviceStatus.history = ServiceStatus.historyFromBackend(status);
        console.log('status.updatedAt - ', status.updatedAt);
        serviceStatus.date = status.updatedAt ? new Date(status.updatedAt) : new Date();
        return serviceStatus;
    }

    private static historyFromBackend(status) {
        const result = [];
        if (status.history) {
            status.history.forEach(h => {
                result.push(ServiceEvent.fromBackend(h, status.components));
            });
        }
        return result;
    }

    dateStr(): string {
        return toLongString(this.date);
    }


    isActiveIssue(): boolean {
        return ACTIVE_STATES.some(s => {
            return s === this.statusType;
        });
    }

    isResolved(): boolean {
        return this.currentEvent.history[this.currentEvent.history.length - 1].type === 'Resolved';
    }

    resolve() {
        if (!this.isResolved() && this.statusType !== 'maintenance') {
            this.currentEvent.history.push(new EventUpdate(new Date(), 'Resolved', 'The issue is now resolved.'));
        }
        this.addToHistory(this.currentEvent);
        this.date = new Date();
        this.statusString = 'Service is operational';
        this.description = 'Welcome to the Service Status Page. There you can see current information of the service performance. ' +
            'You can bookmark or subscribe to this page for the latest updates.';
        this.statusType = 'normal';
        this.components = ServiceStatus.getNormalComponents();
        this.currentEvent = null;
    }

    private addToHistory(event: ServiceEvent) {
        if (this.history.length >= this.MAX_HISTORY_SIZE) {
            this.history = this.history.slice(0, this.MAX_HISTORY_SIZE - 1);
        }
        this.history.unshift(event);
    }
}
